services/rollout-service/pkg/argo/argo.go

```diff
@@ -134,115 +139,163 @@ type ArgoOverview struct {
+		a.KnownApps[envName] = map[string]*v1alpha1.Application{}
 	}
 	switch ev.Type {
 	case "ADDED", "MODIFIED":
 		l.Info("created/updated:kuberpult.application:" + ev.Application.Name + ",kuberpult.environment:" + envName)
-		appsKnownToArgo[envName][appName] = &ev.Application
+		a.KnownApps[envName][appName] = &ev.Application
 	case "DELETED":
 		l.Info("deleted:kuberpult.application:" + ev.Application.Name + ",kuberpult.environment:" + envName)
-		delete(appsKnownToArgo[envName], appName)
+		delete(a.KnownApps[envName], appName)
 	}
 }
 
-func (a *ArgoAppProcessor) CreateOrUpdateApp(ctx context.Context, overview *api.GetOverviewResponse, appName, team string, env *api.Environment, appsKnownToArgo map[string]*v1alpha1.Application) {
-	t := team
+type AppInfo struct {
+	ApplicationName              string
+	TeamName                     string
+	EnvironmentName              string
+	ParentEnvironmentName        string
+	ArgoEnvironmentConfiguration *api.EnvironmentConfig_ArgoCD
+}
 
-	var existingApp *v1alpha1.Application
-	selfManaged, err := IsSelfManagedFilterActive(t, a)
+func (a *ArgoAppProcessor) isKnownArgoApp(appName, envName string, appsKnownToArgo map[string]*v1alpha1.Application) *v1alpha1.Application {
+	for _, argoApp := range appsKnownToArgo {
+		if argoApp.Annotations["com.freiheit.kuberpult/application"] == appName && argoApp.Annotations["com.freiheit.kuberpult/environment"] == envName {
+			return argoApp
+		}
+	}
+	return nil
+}
+
+func (a *ArgoAppProcessor) CreateArgoApp(ctx context.Context, overview *api.GetOverviewResponse, appInfo *AppInfo) {
+	selfManaged, err := IsSelfManagedFilterActive(appInfo.TeamName, a)
 	if err != nil {
 		logger.FromContext(ctx).Error("detecting self manage:", zap.Error(err))
 	}
 	if selfManaged {
-		for _, argoApp := range appsKnownToArgo {
-			if argoApp.Annotations["com.freiheit.kuberpult/application"] == appName && argoApp.Annotations["com.freiheit.kuberpult/environment"] == env.Name {
-				existingApp = argoApp
-				break
+		createSpan, ctx := tracer.StartSpanFromContext(ctx, "CreateApplication")
+		createSpan.SetTag("application", appInfo.ApplicationName)
+		createSpan.SetTag("environment", appInfo.EnvironmentName)
+		createSpan.SetTag("operation", "create")
+		appToCreate := CreateArgoApplication(overview, appInfo)
+		appToCreate.ResourceVersion = ""
+		upsert := false
+		validate := false
+		appCreateRequest := &application.ApplicationCreateRequest{
+			XXX_NoUnkeyedLiteral: struct{}{},
+			XXX_unrecognized:     nil,
+			XXX_sizecache:        0,
+			Application:          appToCreate,
+			Upsert:               &upsert,
+			Validate:             &validate,
+		}
+		_, err := a.ApplicationClient.Create(ctx, appCreateRequest)
+		if err != nil {
+			// We check if the application was created in the meantime
+			if status.Code(err) != codes.InvalidArgument {
+				logger.FromContext(ctx).Error("creating "+appToCreate.Name+",env "+appInfo.EnvironmentName, zap.Error(err))
```

Comment:
```suggestion
				logger.FromContext(ctx).Error("creating argo app, zap.Error(err), zap.Strings("AppName", appToCreate.Name, "EnvName", appInfo.EnvironmentName)
```
Either we use `zap`, then with a fixed error message and all params in a `zap` function (like in my suggestion) - or use Errorf with placeholders like %s/%w to format the string. Please don't mix both approaches.